ipa-core/src/net/test.rs

```diff
@@ -140,63 +353,42 @@ impl TestConfigBuilder {
         self
     }
 
-    #[must_use]
-    pub fn build(self) -> TestConfig {
-        let mut sockets = None;
-        let ports = self.ports.unwrap_or_else(|| {
-            let socks = array::from_fn(|_| TcpListener::bind("localhost:0").unwrap());
-            let ports = socks
-                .each_ref()
-                .map(|sock| sock.local_addr().unwrap().port());
-            sockets = Some(socks);
-            ports
-        });
-        let (scheme, certs) = if self.disable_https {
-            ("http", [None, None, None])
+    /// Creates a HTTP1 or HTTP2 client config.
+    pub fn create_client_config(&self) -> ClientConfig {
+        self.use_http1
+            .then(ClientConfig::use_http1)
+            .unwrap_or_default()
+    }
+
+    /// Get all the MPC ports in a ring specified by the shard index.
+    fn get_ports_for_shard_index(&self, ix: ShardIndex) -> Vec<Option<u16>> {
+        if let Some(ports_by_ring) = &self.ports_by_ring {
+            let ports = ports_by_ring[ix.as_index()].clone();
+            ports.ring.into_iter().map(Some).collect()
         } else {
-            ("https", TEST_CERTS_DER.clone().map(Some))
-        };
-        let peers = certs
-            .into_iter()
-            .enumerate()
-            .map(|(i, cert)| PeerConfig {
-                url: format!("{scheme}://localhost:{}", ports[i])
-                    .parse()
-                    .unwrap(),
-                certificate: cert,
-                hpke_config: if self.disable_matchkey_encryption {
-                    None
-                } else {
-                    Some(HpkeClientConfig::new(
-                        IpaPublicKey::from_bytes(
-                            &hex::decode(TEST_HPKE_PUBLIC_KEY.trim()).unwrap(),
-                        )
-                        .unwrap(),
-                    ))
-                },
-            })
-            .collect::<Vec<_>>();
-        let network = NetworkConfig::<Helper>::new_mpc(
-            peers,
-            self.use_http1
-                .then(ClientConfig::use_http1)
-                .unwrap_or_default(),
-        );
-        let servers = if self.disable_https {
-            ports.map(|ports| server_config_insecure_http(ports, !self.disable_matchkey_encryption))
+            vec![None; 3]
+        }
+    }
+
+    /// Get all the shard ports in a helper.
+    fn get_ports_for_helper_identity(&self, id: HelperIdentity) -> Vec<Option<u16>> {
+        if let Some(ports_by_ring) = &self.ports_by_ring {
+            ports_by_ring
+                .iter()
+                .map(|r| r.shards[id.as_index()])
+                .map(Some)
```

Comment:
```suggestion
                .map(|r| Some(r.shards[id.as_index()]))
```

Reply:
In general, do you prefer larger blocks of code inside a map function instead of more map calls?

Reply:
it is always a balance, cluttered map calls are bad for readability, more map calls result in more complex types used and more work for compiler to unwind them